### PR TITLE
*: drop go 1.18 support, add go 1.21 support

### LINKED ISSUE #2626: Move to using sync/atomic types from Go 1.19
Looks like 1.19 makes `go.uber.org/atomic` somewhat obsolete (at last!): https://go.dev/doc/go1.19#atomic_types

### LINKED ISSUE #2627: Use AppendByteOrder to append integers faster
I doubt it'll make a lot of difference for us and we usually have more problems _prepending_ integers, but still it's worth using with Go 1.19:
https://pkg.go.dev/encoding/binary#AppendByteOrder

### LINKED ISSUE #2629: fmt.Append* for formatted appending to []byte
Can be used since Go 1.19:
https://pkg.go.dev/fmt#Append

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #2626, close #2627, close #2629 and update all workflows/images.

TODO:

- [x] Use flag.TextVar where possible.
- [x] Update interop deps.

### CODE REVIEW

File: pkg/network/fuzz_test.go [hunk 1]
```diff
@@ -1,4 +1,4 @@
-//go:build go1.18
+//go:build go1.19
```
Comment (roman-khimov) [inline pkg/network/fuzz_test.go:1]: These can be dropped completely. They were only relevant pre-1.18.

File: cli/query/query.go [hunk 2]
```diff
@@ -121,46 +121,48 @@ func DumpApplicationLog(
...
 			}
 		}
 	}
+	// Ignore the errors below because `Write` to buffer doesn't return error.
+	resBuf := bytes.NewBuffer(nil)
+	tw := tabwriter.NewWriter(resBuf, 0, 4, 4, '\t', 0)
+	_, _ = tw.Write(buf)
 	_ = tw.Flush()
-	fmt.Fprint(ctx.App.Writer, buf.String())
+	fmt.Fprint(ctx.App.Writer, resBuf.String())
```
Comment (roman-khimov) [inline cli/query/query.go:165]: How about `tabwriter.NewWriter(ctx.App.Writer, ...)` and drop `resBuf`?

**Review by roman-khimov — COMMENTED** [review]
Please check atomic copying as well, it seems to be OK except for the cases noted, but maybe there are other cases like that.

**Comment by AnnaShaleva** [comment]
@roman-khimov, I'm still not sure that it will be useful for us integrate flag.TextVar into our repo. Let's take as an example `flags.Address` and `flags.AddressFlag` structures and `wallet change-password` command that uses `--address` flag:

First of all, we can't completely remove `flags.AddressFlag` structure because `urfave.cli` requires a full set of `cli.Flag` flags to be provided to the command constructor for proper parsing and documentation displaying:
https://github.com/nspcc-dev/neo-go/blob/b284b902d6d6dec3d666e3f140f906e14ff6beb9/cli/wallet/wallet.go#L133-L145

After that, if we're trying to replace `flags.Address` structure by a single `*util.Uint160` value with TextMarshaller interface defined on it, then we can get the following diff:
```
diff --git a/cli/flags/address.go b/cli/flags/address.go
index f20e9948..52736c10 100644
--- a/cli/flags/address.go
+++ b/cli/flags/address.go
@@ -10,53 +10,20 @@ import (
        "github.com/urfave/cli"
 )
 
-// Address is a wrapper for a Uint160 with flag.Value methods.
-type Address struct {
-       IsSet bool
-       Value util.Uint160
-}
-
 // AddressFlag is a flag with type string.
 type AddressFlag struct {
        Name  string
        Usage string
-       Value Address
+       Value *util.Uint160
 }
 
 var (
-       _ flag.Value = (*Address)(nil)
-       _ cli.Flag   = AddressFlag{}
+       _ cli.Flag = AddressFlag{}
 )
 
-// String implements the fmt.Stringer interface.
-func (a Address) String() string {
-       return address.Uint160ToString(a.Value)
-}
-
-// Set implements the flag.Value interface.
-func (a *Address) Set(s string) error {
-       addr, err := ParseAddress(s)
-       if err != nil {
-               return cli.NewExitError(err, 1)
-       }
-       a.IsSet = true
-       a.Value = addr
-       return nil
-}
-
-// Uint160 casts an address to Uint160.
-func (a *Address) Uint160() (u util.Uint160) {
-       if !a.IsSet {
-               // It is a programmer error to call this method without
-               // checking if the value was provided.
-               panic("address was not set")
-       }
-       return a.Value
-}
-
 // IsSet checks if flag was set to a non-default value.
 func (f AddressFlag) IsSet() bool {
-       return f.Value.IsSet
+       return f.Value != nil
 }
 
 // String returns a readable representation of this value
@@ -86,7 +53,7 @@ func (f AddressFlag) GetName() string {
 // Ignores errors.
 func (f AddressFlag) Apply(set *flag.FlagSet) {
        eachName(f.Name, func(name string) {
-               set.Var(&f.Value, name, f.Usage)
+               set.TextVar(f.Value, name, f.Value, f.Usage)
        })
 }
```

It's true that such replacement simplifies our `flags` package code, but at the same time we have to manually perform some utility conversions in the code:
```
diff --git a/cli/wallet/wallet.go b/cli/wallet/wallet.go
index f5051efb..700746fe 100644
--- a/cli/wallet/wallet.go
+++ b/cli/wallet/wallet.go
@@ -355,10 +355,10 @@ func changePassword(ctx *cli.Context) error {
        if len(wall.Accounts) == 0 {
                return cli.NewExitError("wallet has no accounts", 1)
        }
-       addrFlag := ctx.Generic("address").(*flags.Address)
-       if addrFlag.IsSet {
+       addrFlag := ctx.Generic("address").(*util.Uint160)
+       if addrFlag != nil {
                // Check for account presence first before asking for password.
-               acc := wall.GetAccount(addrFlag.Uint160())
+               acc := wall.GetAccount(*addrFlag)
                if acc == nil {
                        return cli.NewExitError("account is missing", 1)
                }
@@ -370,7 +370,7 @@ func changePassword(ctx *cli.Context) error {
        }
 
        for i := range wall.Accounts {
-               if addrFlag.IsSet && wall.Accounts[i].Address != addrFlag.String() {
+               if addrFlag != nil && wall.Accounts[i].Address != address.Uint160ToString(*addrFlag) {
                        continue
                }
                err := wall.Accounts[i].Decrypt(oldPass, wall.Scrypt)

```

So my question is: am I got the idea right, is it the code that we'd like to change? The same logic is related to the Fixed8 flag.

**Comment by roman-khimov** [comment]
@AnnaShaleva, if it doesn't fit well --- forget about it. The idea was to drop these flags, if there are valid reasons why we can't do that then OK, we don't do it. We've got #3097 as well for the future.

File: pkg/services/rpcsrv/server.go [hunk 3]
```diff
@@ -317,7 +317,7 @@ func New(chain Ledger, conf config.RPC, coreServer *network.Server,
 		conf.MaxWebSocketClients = defaultMaxWebSocketClients
 		log.Info("MaxWebSocketClients is not set or wrong, setting default value", zap.Int("MaxWebSocketClients", defaultMaxWebSocketClients))
 	}
-	var oracleWrapped = new(atomic.Value)
+	var oracleWrapped atomic.Value
 	if orc != nil {
 		oracleWrapped.Store(orc)
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/services/rpcsrv/server.go]: Fixed, it's the only case where it's needed to store the pointer.

File: pkg/services/rpcsrv/server_test.go [hunk 4]
```diff
@@ -2333,7 +2333,7 @@ func TestNotaryRequestRPC(t *testing.T) {
 		submitNotaryRequest(t, false, 0, str)(t)
 
 		t.Run("nonempty pool", func(t *testing.T) {
-			//get notary pool & check tx hashes
+			// get notary pool & check tx hashes
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/services/rpcsrv/server_test.go:2336]: Created a separate commit.

File: pkg/core/stateroot/module.go [hunk 5]
```diff
@@ -203,8 +203,9 @@ func (s *Module) Init(height uint32) error {
 // outdated MPT data before state sync process can be started.
 // Note: this method is aimed to be called for genesis block only, an error is returned otherwise.
 func (s *Module) CleanStorage() error {
-	if s.localHeight.Load() != 0 {
-		return fmt.Errorf("can't clean MPT data for non-genesis block: expected local stateroot height 0, got %d", s.localHeight.Load())
+	lH := s.localHeight.Load()
+	if lH != 0 {
+		return fmt.Errorf("can't clean MPT data for non-genesis block: expected local stateroot height 0, got %d", lH)
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/core/stateroot/module.go:208]: Created a separate commit.

File: pkg/network/fuzz_test.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/network/fuzz_test.go]: Fixed.

File: cli/query/query.go (hunk 2 above)
Comment (AnnaShaleva) [reply cli/query/query.go]: Fixed.

File: cli/query/query.go [hunk 6]
```diff
@@ -301,8 +303,10 @@ func queryVoter(ctx *cli.Context) error {
 	if st.VoteTo != nil {
 		voted = fmt.Sprintf("%s (%s)", hex.EncodeToString(st.VoteTo.Bytes()), address.Uint160ToString(st.VoteTo.GetScriptHash()))
 	}
-	fmt.Fprintf(ctx.App.Writer, "\tVoted: %s\n", voted)
-	fmt.Fprintf(ctx.App.Writer, "\tAmount : %s\n", fixedn.ToString(&st.Balance, int(dec)))
-	fmt.Fprintf(ctx.App.Writer, "\tBlock: %d\n", st.BalanceHeight)
```
Comment (roman-khimov) [inline cli/query/query.go:306]: Why changing this? It's a pretty direct write that doesn't require additional buffering.

File: cli/query/query.go [hunk 7]
```diff
@@ -196,15 +196,17 @@ func queryCandidates(ctx *cli.Context) error {
...
 	for _, val := range vals {
-		_, _ = tw.Write([]byte(fmt.Sprintf("%s\t%d\t%t\t%t\n", hex.EncodeToString(val.PublicKey.Bytes()), val.Votes, comm.Contains(&val.PublicKey), val.Active)))
+		res = fmt.Appendf(res, "%s\t%d\t%t\t%t\n", hex.EncodeToString(val.PublicKey.Bytes()), val.Votes, comm.Contains(&val.PublicKey), val.Active)
 	}
-	_ = tw.Flush()
-	fmt.Fprint(ctx.App.Writer, buf.String())
-	return nil
+	tw := tabwriter.NewWriter(ctx.App.Writer, 0, 2, 2, ' ', 0)
+	_, err = tw.Write(res)
+	if err != nil {
```
Comment (roman-khimov) [inline cli/query/query.go:206]: But you check for error here.

File: cli/query/query.go [hunk 8]
```diff
@@ -121,46 +121,46 @@ func DumpApplicationLog(
...
 		if res != nil {
 			for _, e := range res.Executions {
 				if e.VMState != vmstate.Halt {
-					_, _ = tw.Write([]byte("Exception:\t" + e.FaultException + "\n"))
+					buf = fmt.Appendf(buf, "Exception:\t%s\n", e.FaultException)
 				}
 			}
 		}
 	}
+	// Ignore the errors below because `Write` to buffer doesn't return error.
```
Comment (roman-khimov) [inline cli/query/query.go:160]: While ignoring here.

**Review by roman-khimov — COMMENTED** [review]
563765b462e1d2d61b1f6fa66e9d4fb537bd90a7 has "Close 2626", should be `#2626` at least.

File: cli/query/query.go (hunk 8 above)
Comment (AnnaShaleva) [reply cli/query/query.go]: Should be checked here as far, fixed.

File: cli/query/query.go (hunk 6 above)
Comment (AnnaShaleva) [reply cli/query/query.go]: Removed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**